keanu-project/src/main/java/io/improbable/keanu/vertices/Vertex.java

```diff
@@ -98,7 +98,35 @@ public double logDensityAtValue() {
      * @return The value at this vertex after recalculating any parent non-probabilistic
      * vertices.
      */
-    public abstract T lazyEval();
+    public T lazyEval() {
+        Deque<Vertex<?>> stack = new ArrayDeque<>();
+        stack.push(this);
+        Set<Vertex<?>> hasCalculated = new HashSet<>();
+
+        while (!stack.isEmpty()) {
+
+            Vertex<?> head = stack.peek();
+            Set<Vertex<?>> parents = head.getParents();
+
+            if (parents.isEmpty() || areParentsCalculated(hasCalculated, parents) || head.isProbabilistic()) {
```

Comment:
Just realized that this iterates through the parents twice. Once in areParentsCalculated and then in the else branch. Ideally, the areParentsCalculated method would return the parents that aren't calculated instead of just a boolean that says there are un-calculated parents.

Reply:
Nice call! I'll put that in